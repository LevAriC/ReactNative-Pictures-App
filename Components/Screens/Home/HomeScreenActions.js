import { TOGGLE_VIEW, SEARCH_PICS, UPDATE_SEARCH_QUERY, SAVE_LAST_VIEW, UPDATE_LOADING } from './HomeScreenActionTypes'

const toggleView = (view) => ({
    type: TOGGLE_VIEW,
    data: { view }
})
const handleToggleView = (view) => async dispatch => {
    dispatch(toggleView(view))
}

const updateIsLoading = (isLoading) => ({
    type: UPDATE_LOADING,
    data: { isLoading }
})
const handleUpdateIsLoading = (isLoading) => async dispatch => {
    dispatch(updateIsLoading(isLoading))
}

const saveLastView = (view) => ({
    type: SAVE_LAST_VIEW,
    data: { view }
})
const handleSaveLastView = (view) => async dispatch => {
    dispatch(saveLastView(view))
}

const updateSearchQuery = (searchQuery) => ({
    type: UPDATE_SEARCH_QUERY,
    data: { searchQuery }
})
const handleUpdateSearchQuery = (searchQuery) => async dispatch => {
    dispatch(updateSearchQuery(searchQuery))
}

const updatePicsList = (picList) => ({
    type: SEARCH_PICS,
    data: { picList }
})
const handlePicsSearch = (searchQuery) => async dispatch => {
    const pixabayKey = '12263176-901abf3c360dd1fbe2da948f7'
    const url = `https://pixabay.com/api/?key=${pixabayKey}&q=${searchQuery}&image_type=photo&per_page=198`
    const options = { method: 'GET', headers: { 'Content-Type': 'application/json', 'Accept':'application/json' } }
    const request = new Request(url, options)

    await fetch(request)
    .then(response => response.json())
    .then(async data => dispatch(updatePicsList(data.hits)))
    .catch(err => console.log(err))
}

export default {
    handlePicsSearch,
    handleToggleView,
    updatePicsList,
    toggleView,
    updateSearchQuery,
    handleUpdateSearchQuery,
    saveLastView,
    handleSaveLastView,
    updateIsLoading,
    handleUpdateIsLoading
}