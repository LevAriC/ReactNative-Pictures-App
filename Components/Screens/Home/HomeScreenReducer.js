import { TOGGLE_VIEW, SEARCH_PICS, SAVE_LAST_VIEW, UPDATE_LOADING, SAVE_LAST_LIST } from './HomeScreenActionTypes'

const initialState = {
    view: 'Grid',
    lastView: '',
    picsList: [],
    lastPicsList: [],
    isLoading: true
}

export default (state = initialState, action) => {
    switch (action.type) {
        case SEARCH_PICS:
            return {
                ...state,
                picsList: action.data.picList
            }
        case TOGGLE_VIEW:
            return {
                ...state,
                view: action.data.view
            }
        case SAVE_LAST_VIEW:
            return {
                ...state,
                lastView: action.data.view
            }
        case UPDATE_LOADING:
            return {
                ...state,
                isLoading: action.data.isLoading
            }   
        case SAVE_LAST_LIST:
            return {
                ...state,
                lastPicsList: action.data.picsList
            }  
        default:
            return state
    }
}