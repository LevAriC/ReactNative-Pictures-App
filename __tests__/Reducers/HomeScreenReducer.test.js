import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import rootReducer from '../../Components/rootReducer'
import HomeScreenActions from '../../Components/Screens/Home/HomeScreenActions'
import HomeScreenReducer from '../../Components/Screens/Home/HomeScreenReducer'

const middleware = applyMiddleware(thunk)
const composedEnhancers = compose(middleware)
const initialState = {}

it('HomeScreen - Initial State', () => {
    const store = createStore(rootReducer, initialState, composedEnhancers)
    expect(store.getState().HomeScreen.view).toMatchSnapshot()
    expect(store.getState().HomeScreen.lastView).toMatchSnapshot()
    expect(store.getState().HomeScreen.picsList).toMatchSnapshot()
    expect(store.getState().HomeScreen.lastPicsList).toMatchSnapshot()
    expect(store.getState().HomeScreen.isLoading).toMatchSnapshot()
});

it('HomeScreen - Picture Search', () => {
  const store = createStore(rootReducer, initialState, composedEnhancers)
  store.dispatch(HomeScreenActions.updatePicsList({ Yoda: 'YES' }))
  expect(store.getState().HomeScreen.picsList).toHaveProperty('Yoda', 'YES')
  store.dispatch(HomeScreenActions.handleSaveLastPicsList(store.getState().HomeScreen.picsList))
  expect(store.getState().HomeScreen.lastPicsList).toHaveProperty('Yoda', 'YES')    
});

it('HomeScreen - Layout Toggle', () => {
    const store = createStore(rootReducer, initialState, composedEnhancers)
    expect(store.getState().HomeScreen.view).toMatchSnapshot()
    store.dispatch(HomeScreenActions.handleToggleView("List"))
    expect(store.getState().HomeScreen.view).toMatchSnapshot()
    store.dispatch(HomeScreenActions.handleToggleView("Grid"))
    expect(store.getState().HomeScreen.view).toMatchSnapshot()
    store.dispatch(HomeScreenActions.handleToggleView("Picture"))
    expect(store.getState().HomeScreen.view).toMatchSnapshot()
    store.dispatch(HomeScreenActions.handleSaveLastView("Picture"))
    expect(store.getState().HomeScreen.lastView).toMatchSnapshot()
});