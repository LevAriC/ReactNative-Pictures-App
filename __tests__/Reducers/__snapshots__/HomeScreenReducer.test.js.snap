// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen - Initial State 1`] = `"Grid"`;

exports[`HomeScreen - Initial State 2`] = `""`;

exports[`HomeScreen - Initial State 3`] = `Array []`;

exports[`HomeScreen - Initial State 4`] = `Array []`;

exports[`HomeScreen - Initial State 5`] = `true`;

exports[`HomeScreen - Layout Toggle 1`] = `"Grid"`;

exports[`HomeScreen - Layout Toggle 2`] = `"List"`;

exports[`HomeScreen - Layout Toggle 3`] = `"Grid"`;

exports[`HomeScreen - Layout Toggle 4`] = `"Picture"`;

exports[`HomeScreen - Layout Toggle 5`] = `"Picture"`;
